### PR TITLE
Add support for Federated Auth to test resources scripts

### PR DESCRIPTION
**Pull request by danieljurek**
Fiixes #8245

### CODE REVIEW

File: eng/common/TestResources/New-TestResources.ps1 [hunk 1]
```diff
@@ -92,8 +92,13 @@ param (
     [Parameter()]
     [switch] $SuppressVsoCommands = ($null -eq $env:SYSTEM_TEAMPROJECTID),
 
+    # Default behavior is to use -UserAuth
     [Parameter()]
-    [switch] $UserAuth,
+    [switch] $ServicePrincipalAuth,
+
+    [Parameter()]
+    [switch] $FederatedAuth,
```
Comment (benbp) [inline eng/common/TestResources/New-TestResources.ps1:100]: Do we want to throw this under the Provisioner parameter set?

File: eng/common/TestResources/New-TestResources.ps1 [hunk 2]
```diff
@@ -268,7 +283,6 @@ function BuildDeploymentOutputs([string]$serviceName, [object]$azContext, [objec
     # Add default values
     $deploymentOutputs = [Ordered]@{
         "${serviceDirectoryPrefix}CLIENT_ID" = $TestApplicationId;
-        "${serviceDirectoryPrefix}CLIENT_SECRET" = $TestApplicationSecret;
         "${serviceDirectoryPrefix}TENANT_ID" = $azContext.Tenant.Id;
```
Comment (benbp) [inline eng/common/TestResources/New-TestResources.ps1:286]: Do we want to remove all three of these vars from being set?

File: eng/common/TestResources/deploy-test-resources.yml [hunk 3]
```diff
@@ -4,7 +4,10 @@ parameters:
   DeleteAfterHours: 8
   Location: ''
   SubscriptionConfiguration: $(sub-config-azure-cloud-test-resources)
+  ServiceConnection: not-specified
   ResourceType: test
+  UseFederatedAuth: true
```
Comment (benbp) [inline eng/common/TestResources/deploy-test-resources.yml:9]: I think we'll want this to default to false to start?

**Review by benbp — COMMENTED** [review]
A couple comments but looking good. @weshaggard to take a look.

File: eng/common/TestResources/deploy-test-resources.yml (hunk 3 above)
Comment (danieljurek) [reply eng/common/TestResources/deploy-test-resources.yml:9]: I think it should be on by default and turned it off in `tests.yml` for services which cannot migrate.

We'd start with your inventory of services and if service owners spot failures they can turn it off themselves.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 1 above)
Comment (danieljurek) [reply eng/common/TestResources/New-TestResources.ps1:100]: If this was placed in a parameter set then it'll require that the other mandatory parameters be specified. In that case, one will always have to specify the provisioner parameters. I think this should be outside the Provisioner param set.

File: eng/common/TestResources/deploy-test-resources.yml (hunk 3 above)
Comment (benbp) [reply eng/common/TestResources/deploy-test-resources.yml:9]: Eventually, but it means that we basically can't merge this in to the repos right now without coordinating a massive yaml change. If we set it to false then we can start to migrate one by one, and then once we reach a migrated state with exceptions, we can invert the parameter.

**Comment by joshfree** [comment]
+ @KarishmaGhiya too

File: eng/common/TestResources/deploy-test-resources.yml [hunk 4]
```diff
@@ -38,26 +42,33 @@ steps:
...
-        -AdditionalParameters ${{ parameters.ArmTemplateParameters }} `
-        -CI `
-        -Force `
-        -Verbose | Out-Null
+  - task: AzurePowerShell@5
     displayName: Deploy test resources
     env:
       TEMP: $(Agent.TempDirectory)
+    inputs:
+      azureSubscription: ${{ parameters.ServiceConnection }}
```
Comment (danieljurek) [inline eng/common/TestResources/deploy-test-resources.yml:50]: TODO: only use `azureSubscription` if `ServiceConnection` is specified.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 1 above)
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:100]: Is "FederatedAuth" the correct term here? It seems like that might be one form of logged in user but not all. Would "UseLoggedInCredentials" or something like that be better?

File: eng/common/TestResources/deploy-test-resources.yml [hunk 5]
```diff
@@ -38,26 +42,33 @@ steps:
...
+      azurePowerShellVersion: LatestVersion
+      pwsh: true
+      ScriptType: InlineScript
+      Inline: |
+        eng/common/scripts/Import-AzModules.ps1
+
+        $subscriptionConfiguration = @'
+          ${{ parameters.SubscriptionConfiguration }}
```
Comment (weshaggard) [inline eng/common/TestResources/deploy-test-resources.yml:58]: Do we need to delete or remove some of the existing SP credentials from the configurations?

File: eng/common/TestResources/deploy-test-resources.yml (hunk 4 above)
Comment (weshaggard) [reply eng/common/TestResources/deploy-test-resources.yml:50]: If that is empty won't the AzurePowerShell step fail?

File: eng/common/TestResources/New-TestResources.ps1 [hunk 6]
```diff
@@ -625,8 +651,8 @@ try {
         $userAccountName = $userAccount.UserPrincipalName
         Log "User authentication with user '$userAccountName' ('$TestApplicationId') will be used."
     }
-    # If no test application ID was specified during an interactive session, create a new service principal.
-    elseif (!$CI -and !$TestApplicationId) {
+    # If user has specified -ServicePrincipalAuth
+    elseif (!$CI -and $ServicePrincipalAuth) {
```
Comment (heaths) [inline eng/common/TestResources/New-TestResources.ps1:655]: Seems this needs to require `$TestApplicationId`, then. PowerShell does support a rich set of APIs for validating parameters if you want to know more. It's more idiomatic and does positively affect the CLI.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 1 above)
Comment (danieljurek) [reply eng/common/TestResources/New-TestResources.ps1:100]: `-UserAuth` already uses logged in credentials but it does other things like resolve a service principal ID from the logged in user and makes that the test app. 

Another approach could be to change `-UserAuth` to `-UseLoggedInCredential` and decide whether to resolve the service principal ID based on whether `-CI` is set.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 6 above)
Comment (danieljurek) [reply eng/common/TestResources/New-TestResources.ps1:655]: We changed behavior in the script to default to the `-UserAuth` behavior of today. `-ServicePrinipalAuth` creates a service principal on behalf of the user for testing. 

Perhaps this should be something more like `-CreateServicePrincipal`

Either way `TestApplicationId` would be ignored.

File: eng/common/TestResources/deploy-test-resources.yml (hunk 5 above)
Comment (danieljurek) [reply eng/common/TestResources/deploy-test-resources.yml:81]: In the case of `-UseFederatedAuth` any secrets passed to the script are NOT passed to the templates or pre-/post- scripts. 

This way we don't have to make changes to sub-configs until all services have migrated.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 1 above)
Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:100]: I think we could just remove `FederatedAuth` entirely. If `ServicePrincipalAuth` is not set, then we look for logged in credentials via azure powershell/azure cli login context (or try running Connect-AzAccount in the non-ci context). Then the federated auth parameter you have in the pipeline can just default to `-ServicePrincipalAuth:$${{ parameters.UseFederatedAuth }}`.

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:100]: So yes, CI and/or ServicePrincipalAuth parameters are what sets the mode/behavior.

File: eng/common/TestResources/deploy-test-resources.yml (hunk 5 above)
Comment (benbp) [reply eng/common/TestResources/deploy-test-resources.yml:81]: Yes the plan here was not to change the sub configs or introduce layers as its a larger change. Trying to keep each incremental step small here so we can start unblocking sdk teams who have DAC changes on the way.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 6 above)
Comment (danieljurek) [reply eng/common/TestResources/New-TestResources.ps1:640]: We kept the name `-ServicePrincipalAuth` but changed the behaviors a bit. We now use `ServicePrincipalAuth` to determine whether to log in as the provisioner app optionally supplied in `-ProvisionerApplication___` or to create a service principal on behalf of the user if no provisioner app info is supplied. 

The `-UserAuth` param was removed and it's now the default behavior.

**Review by benbp — APPROVED** [review]

**Review by weshaggard — APPROVED** [review]

**Pull request merged**